internals/overlord/logstate/gatherer.go

```diff
@@ -182,17 +187,37 @@ func (g *logGatherer) PlanChanged(pl *plan.Plan, serviceData map[string]*Service
 
 		g.pullers.Add(service.Name, svcData.Buffer, g.entryCh)
 		// update labels
-		g.setLabels <- svcWithLabels{service.Name, labels[service.Name]}
+		if dying := g.setClientLabels(service.Name, labels[service.Name]); dying {
+			return
+		}
 	}
 }
 
 // ServiceStarted is called by the LogManager on the start of a service which
 // logs to this gatherer's target.
+// The labels map should not be modified while this method is running.
 func (g *logGatherer) ServiceStarted(service *plan.Service, buffer *servicelog.RingBuffer, labels map[string]string) {
 	g.pullers.Add(service.Name, buffer, g.entryCh)
 
 	// Add this service's custom labels to the client
-	g.setLabels <- svcWithLabels{service.Name, labels}
+	if dying := g.setClientLabels(service.Name, labels); dying {
+		return
+	}
+}
+
+// setClientLabels tells the main loop to set the labels on the client. This
+// method is safe to use concurrently.
+// It also checks if the gatherer is dying - in which case, it will return
+// dying == true to signal to the caller that they should return.
+func (g *logGatherer) setClientLabels(serviceName string, labels map[string]string) (dying bool) {
```

Comment:
I don't think returning true or false here makes much sense. It isn't the responsibility of this set function to determine if we are dying. We just don't want this send here to block, hence the need for the select.

Handling dying is the concern of the caller.

Reply:
the whole point of this function was to factor out common logic
```go
	select {
	case <-g.tomb.Dying():
		return
	case g.setLabels <- svcWithLabels{serviceName, labels}:
	}
```

but maybe it's not really helping and I should just copy-paste that block in the code?

Reply:
Yeah, I think that'd probably be better. Couple more lines of code, but the flow would be very clear.